staging/src/k8s.io/apimachinery/pkg/api/meta/help.go

```diff
@@ -143,6 +143,67 @@ func EachListItem(obj runtime.Object, fn func(runtime.Object) error) error {
 		if takeAddr {
 			raw = raw.Addr()
 		}
+		// raw must be a pointer
+		// allocate a pointer is cheap
+		switch item := raw.Interface().(type) {
+		case *runtime.RawExtension:
+			if err := fn(item.Object); err != nil {
+				return err
+			}
+		case runtime.Object:
+			if err := fn(item); err != nil {
+				return err
+			}
+		default:
+			obj, ok := item.(runtime.Object)
+			if !ok {
+				return fmt.Errorf("%v: item[%v]: Expected object, got %#v(%s)", obj, i, raw.Interface(), raw.Kind())
+			}
+			if err := fn(obj); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// EachListItemWithAlloc invokes fn on each runtime.Object in the list.
+// the difference from  EachListItem is that an object of the same type is regenerated and assigned a value,
+// so that the returned Object list and obj have independent life cycles.
+func EachListItemWithAlloc(obj runtime.Object, fn func(runtime.Object) error) error {
```

Comment:
Can you make both this function and EachListItem call a helper function with a flag instead of repeating all the code? That way it will be clear how the functionality differs (and the github diff will be much more understandable).

Reply:
big +1, an example of that (without your more recent type comparison fixes) is at https://github.com/liggitt/kubernetes/commits/ftr/extract_listobject_use_copy

Reply:
thanks for the review suggestion, it has been modified. PTAL...